### PR TITLE
Indexing rewrite

### PR DESCRIPTION
**Pull request by msarahan**
Most of this is adapted from https://github.com/ContinuumIO/conda-channel-clone/ (private project, sorry for curious outsiders)

### CODE REVIEW

**Comment by mikecormier** [comment]
Hey @msarahan, could you add this to your pull request? https://github.com/conda/conda/issues/7583#issuecomment-415400776

**Comment by msarahan** [comment]
Done.  Please test it!

File: conda_build/build.py [hunk 1]
```diff
@@ -1846,16 +1825,6 @@ def test(recipedir_or_package_or_metadata, config, stats, move_broken=True):
     utils.rm_rf(metadata.config.test_dir)
     _extract_test_files_from_package(metadata)
 
-    # When testing a .tar.bz2 in the pkgs dir, clean_pkg_cache() will remove it.
-    # Prevent this. When https://github.com/conda/conda/issues/5708 gets fixed
-    # I think we can remove this call to clean_pkg_cache().
-    in_pkg_cache = (not hasattr(recipedir_or_package_or_metadata, 'config') and
```
Comment (msarahan) [inline conda_build/build.py]: This provides a nice speedup, especially on conda 4.5, but unfortunately, it's still not well behaved in conda (tested on 4.5.5).  This bit needs to stay.

**Comment by msarahan** [comment]
@mikecormier last chance.  If you can't test this to verify that it works now, I'll go ahead anyway, but I'll move on to other things and may not be able to return to fix this for a while.

**Pull request merged**